elastic/kibana
# [Synthetics] add project monitor GET route

## Summary

Relates to https://github.com/elastic/kibana/issues/142418

Adds a route to fetch all monitors for a given project.

`
GET /api/synthetics/project/{projectName}/monitors
`

Returns the following schema, each object of the following schema:
```
{
   total: 600 // the total number of monitors for the project
   /* the key to pass to `search_after` for requesting the next page
    * it is possible for ES to return an array for this value, so the value may be
    * a comma separated list */
   after_key: 'test-monitor', 
   monitors: [
     {
         // the journey id for the monitor
         "journey_id": "test-journey-id", 
          // the config hash for the monitor. Returns "" when not available
         "hash": "test-config-hash",
         /* the sort value returned by the saved object api. This sort value is meant to be used as a `searchAfter` 
          * key. The client should return all values from sort as a comma separated list under `after_id` 
          * when requesting the next page */
         "sort": ["test-journey-id"],
      }
   ]
}
```

### Testing

Create a group of project monitors as normal
1. Navigate to Kibana. Generate an api key
2. Create a new project using the synthetics-agent, or reuse a new old one
3. Navigate to that project, run `SYNTHETICS_API_KEY={theykey} npm run push`
4. Query new api to fetch all given monitors for that project via `curl -X GET http://{username}:{password}@{basepath}/api/synthetics/project/{projectName}/monitors -H 'kbn-xsrf: true'` Make sure to url-encode your project name
5. You should see a response with your monitors. Example
```
{
   total: 1 // the total number of monitors for the project
   /* the key to pass to `search_after` for requesting the next page
    * it is possible for ES to return an array for this value, so the value may be
    * a comma separated list */
   after_key: 'test-journey-id', 
   monitors: [
     {
         // the journey id for the monitor
         "journey_id": "test-journey-id", 
          // the config hash for the monitor. Returns "" when not available
         "hash": "test-config-hash",
         /* the sort value returned by the saved object api. This sort value is meant to be used as a `searchAfter` 
          * key. The client should return all values from sort as a comma separated list under `after_id` 
          * when requesting the next page */
         "sort": ["test-journey-id"],
      }
   ]
}
```

To test pagination, you can create pass the `per_page` key as a parameter. To get to the next page for a given `per_page`, pass the `after_key` via the `search_after` query. 
Example: `curl -X GET http://{username}:{password}@{basepath}/api/synthetics/project/{projectName}/monitors?per_page=10&search_after=test%20journey%20id" -H 'kbn-xsrf: true'`
